contracts/ERC20MultiDelegate.sol

```diff
@@ -173,9 +186,72 @@ contract ERC20MultiDelegate is ERC1155, Ownable {
         require(token.transferFrom(proxyAddressFrom, msg.sender, amount));
     }
 
-    function setUri(string memory uri) external onlyOwner {
-        _setURI(uri);
-        emit MetadataURIUpdated(uri);
+    /**
+     * @dev Generates an onchain metadata for a given tokenId.
+     *
+     * @param tokenId The token ID (address) of the delegate.
+     * @return Onchain metadata in base64 format "data:application/json;base64,<encoded-json>".
+     */
+    function tokenURI(uint256 tokenId) public view returns (string memory) {
+        // convert tokenId to a hex string representation of the address
+        string memory hexAddress = address(uint160(tokenId)).addressToHex();
+
+        // construct the encoded reversed name
+        bytes memory encodedReversedName = bytes.concat(
+            "\x28",
+            bytes(hexAddress),
+            "\x04addr\x07reverse\x00"
+        );
+
+        string memory resolvedName;
+        // attempt to resolve the reversed name using the metadataResolver
+        try metadataResolver.reverse(encodedReversedName) returns (
+            string memory _resolvedName,
+            address,
+            address,
+            address
+        ) {
+            resolvedName = _resolvedName;
+        } catch {}
+
+        string memory imageUri = "";
+
+        if (bytes(resolvedName).length > 0) {
+            (bytes memory encodedName, bytes32 namehash) = resolvedName
+                .dnsEncodeName();
+            bytes memory data = abi.encodeWithSignature(
+                "text(bytes32,string)",
+                [namehash, "avatar"]
+            );
+
+            // attempt to resolve the avatar using the universal resolver
+            try metadataResolver.resolve(encodedName, data) returns (
+                bytes memory _imageUri,
+                address
+            ) {
+                imageUri = _imageUri.length == 0
+                    ? ""
+                    : abi.decode(_imageUri, (string));
+            } catch {}
+        } else {
+            resolvedName = hexAddress;
+        }
+
+        string memory json = Base64.encode(
+            bytes(
+                string.concat(
+                    '{"name": "',
+                    resolvedName,
```

Comment:
This and `imageUri` need to be escaped.

Reply:
Do you think double quotes and backslash escaping will be enough?